File: app/Models/Certificate.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\SoftDeletes;
use Illuminate\Database\Eloquent\Relations\BelongsTo;

class Certificate extends Model
{
    use SoftDeletes;

    protected $fillable = [
        'template_id',
        'student_id',
        'template_recipient_group_id',
        'release_date',
        'verification_code',
    ];

    protected $casts = [
        'release_date' => 'date',
    ];

    /**
     * Get the template associated with this certificate
     */
    public function template(): BelongsTo
    {
        return $this->belongsTo(Template::class);
    }

    /**
     * Get the student who owns this certificate
     */
    public function student(): BelongsTo
    {
        return $this->belongsTo(Student::class);
    }

    /**
     * Get the recipient group associated with this certificate
     */
    public function recipientGroup(): BelongsTo
    {
        return $this->belongsTo(TemplateRecipientGroup::class, 'template_recipient_group_id');
    }
}

File: src/main/kotlin/models/Certificate.kt
package models

import com.expediagroup.graphql.server.extensions.getValueFromDataLoader
import graphql.schema.DataFetchingEnvironment
import kotlinx.serialization.Serializable
import kotlinx.datetime.LocalDate
import kotlinx.datetime.LocalDateTime
import schema.dataloaders.StudentDataLoader
import schema.dataloaders.TemplateDataLoader
import java.util.concurrent.CompletableFuture

@Serializable
data class Certificate(
    val id: Int = 0,
    val templateId: Int,
    val studentId: Int,
    val templateRecipientGroupId: Int,
    val releaseDate: LocalDate,
    val verificationCode: String,
    val createdAt: LocalDateTime,
    val updatedAt: LocalDateTime
){
    fun student(dataFetchingEnvironment: DataFetchingEnvironment): CompletableFuture<Student?> {
        return dataFetchingEnvironment.getValueFromDataLoader(StudentDataLoader.dataLoaderName, studentId)
    }

    fun template(dataFetchingEnvironment: DataFetchingEnvironment): CompletableFuture<Template?> {
        return dataFetchingEnvironment.getValueFromDataLoader(TemplateDataLoader.dataLoaderName, templateId)
    }
}
